src/Hopper/Internal/LoweredCore/ANF.hs

```diff
@@ -0,0 +1,477 @@
+-- | The former with open binders increased for each extra binder present
+-- in the latter. Assumes that the latter is the former extended with extra top
+-- 'BindingLevel's.
+withBinderIncreasesPer :: BindingStack -> BindingStack -> BindingStack
+withBinderIncreasesPer base extended =
+  base & _head.levelIntros                            +~ numBinders
+       & _head.levelRefs.mapped.localNameless.lnDepth +~ numBinders
+  where
+    numBinders = extended `bindersAddedSince` base
+
+-- | Converts a 'Term' in nontail position to ANF. This function corresponds to
+-- Danvy's function "E_c".
+convertWithCont :: Term
+                -- ^ The term to be lowered
+                -> Binding
+                -- ^ The binding to be used, with which future computation will
+                -- refer to the result
+                -> (StackTransform -> LoweringM Anf)
+                -- ^ The continued lowering of the rest of the program, awaiting
+                -- a transformation to update the 'BindingStack'.
+                -> LoweringM Anf
+                -- ^ The action which produces lowered ANF for this term
+convertWithCont term binding k = case term of
+  V v -> do
+    translatedVar <- reader $ translateTermVar v
+    k $ trackVariables binding [translatedVar]
+
+  BinderLevelShiftUP _ _ ->
+    -- TODO: expose this using proper error machinery
+    error "unexpected binder shift in convertWithCont during ANF conversion"
+
+  ELit l -> do
+    body <- k $ trackBinding binding
+    return $ AnfLet (Arity 1)
+                    (RhsAlloc $ AllocLit l)
+                    body
+
+  Return terms ->
+    convertToVars (V.toList terms) $ \vars cleanupRefs ->
+      case binding of
+        TermBinding ->
+          k $ trackVariables binding vars . cleanupRefs
+        (AnfBinding _) ->
+          -- TODO: expose this using proper error machinery
+          error "unexpected non-tail Return outside of a Let RHS"
+
+  EnterThunk t ->
+    convertToVars [t] $ \[var] cleanupRef -> do
+      body <- k $ trackBinding binding . cleanupRef
+      return $ AnfLet (Arity 1) -- TODO: support tupled return
+                      (RhsApp $ AppThunk var)
+                      body
+
+  Delay t -> do
+    thunkBody <- convertGuarded t
+    letBody <- k $ trackBinding binding
+    return $ AnfLet (Arity 1)
+                    (RhsAlloc $ AllocThunk thunkBody)
+                    letBody
+
+  App ft ats -> do
+    let terms = ft : V.toList ats
+    convertToVars terms $ \vars cleanupRefs -> do
+      body <- k $ trackBinding binding . cleanupRefs
+      return $ AnfLet (Arity 1) -- TODO: support tupled return
+                      (RhsApp $ AppFun (head vars)
+                                       (V.fromList $ tail vars))
+                      body
+
+  PrimApp primId terms ->
+    convertToVars (V.toList terms) $ \vars cleanupRefs -> do
+      body <- k $ trackBinding binding . cleanupRefs
+      return $ AnfLet (Arity 1) -- TODO: support tupled return
```

Comment:
At Khan Academy it was conventional to tag TODOs with the owner's name, which I would like to do in Hopper, if you don't mind.